dhcpv6/async.go

```diff
@@ -0,0 +1,214 @@
+package dhcpv6
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"sync"
+	"time"
+)
+
+// ClientAsync implements an asynchronous DHCPv6 client
+type ClientAsync struct {
+	ReadTimeout  time.Duration
+	WriteTimeout time.Duration
+	LocalAddr    net.Addr
+	RemoteAddr   net.Addr
+
+	connection   *net.UDPConn
+	cancel       context.CancelFunc
+	stopping     chan bool
+	receiveQueue chan DHCPv6
+	sendQueue    chan DHCPv6
+	packetsLock  sync.Mutex
+	packets      map[uint32](chan Response)
+}
+
+// NewClientAsync creates an asynchronous client
+func NewClientAsync() *ClientAsync {
+	return &ClientAsync{
+		ReadTimeout:  DefaultReadTimeout,
+		WriteTimeout: DefaultWriteTimeout,
+	}
+}
+
+// OpenForInterface starts the client on the specified interface
+func (c *ClientAsync) OpenForInterface(ifname string, bufferSize int) error {
+	addr, err := GetLinkLocalAddr(ifname)
+	if err != nil {
+		return err
+	}
+	c.LocalAddr = &net.UDPAddr{IP: *addr, Port: DefaultClientPort, Zone: ifname}
+	return c.Open(bufferSize)
+}
+
+// Open starts the client
+func (c *ClientAsync) Open(bufferSize int) error {
+	var (
+		addr *net.UDPAddr
+		ok   bool
+		err  error
+	)
+
+	if addr, ok = c.LocalAddr.(*net.UDPAddr); !ok {
+		return fmt.Errorf("Invalid local address: %v not a net.UDPAddr", c.LocalAddr)
+	}
+
+	// prepare the socket to listen on for replies
+	c.connection, err = net.ListenUDP("udp6", addr)
+	if err != nil {
+		return err
+	}
+	c.stopping = make(chan bool)
+	c.sendQueue = make(chan DHCPv6, bufferSize)
+	c.receiveQueue = make(chan DHCPv6, bufferSize)
+	c.packets = make(map[uint32](chan Response))
+	c.packetsLock = sync.Mutex{}
+
+	var ctx context.Context
+	ctx, c.cancel = context.WithCancel(context.Background())
+	go c.listen(ctx)
+	go c.broadcast(ctx)
+
+	return nil
+}
+
+// Close stops the client
+func (c *ClientAsync) Close() {
+	c.cancel()
+	<-c.stopping
+	<-c.stopping
+	close(c.sendQueue)
+	close(c.receiveQueue)
+	c.connection.Close()
+}
+
+func (c *ClientAsync) listen(ctx context.Context) {
+	defer func() { c.stopping <- true }()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case packet := <-c.receiveQueue:
+			c.receive(packet)
+		}
+	}
+}
+
+func (c *ClientAsync) broadcast(ctx context.Context) {
```

Comment:
this function should be called differently, there is no broadcasting involved (and if we really want, at most it's multicasting). I think a more appropriate name could be `senderLoop`. Similarly `listen` above should be named `receiverLoop` or similar